logo.js

```diff
@@ -841,19 +911,78 @@ function LogoInterpreter(turtle, stream, savehook)
       var atoms = parse(string);
 
       // And execute it!
-      return self.execute(atoms, options);
+      var promise = self.execute(atoms, options).then(null, function (err) {
+        if (err && err.special == "bye") {
+          return undefined;
+        }
+        if (self.turtle) {
+          self.turtle.end();
+        }
+        throw err;
+      });
+      this._setPendingPromise(promise);
+      self._pendingPromise = promise;
+      return promise;
     } catch (e) {
+      // FIXME: doesn't work with promises
```

Comment:
Can you verify/explain? It "should" (in that "computers always make sense" sort of way...)